File: design_patterns/07_Iterator/books.py
class Book:
    def __init__(self, title, author, year):
        self.title = title
        self.author = author
        self.year = year

    def __str__(self):
        return f'{self.title} by {self.author}, {self.year}'

class Library:
    def __init__(self):
        self.books = []

    def add_book(self, book):
        self.books.append(book)

    def __iter__(self):
        return LibraryIterator(self.books)

    def get_books_by_author(self, author):
        return LibraryAuthorIterator(self.books, author)

    def get_books_by_year(self, year):
        return LibraryYearIterator(self.books, year)

class LibraryIterator:
    def __init__(self, books):
        self._books = books
        self._index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self._index < len(self._books):
            result = self._books[self._index]
            self._index += 1
            return result
        else:
            raise StopIteration

class LibraryAuthorIterator:
    def __init__(self, books, author):
        self._books = [book for book in books if book.author == author]
        self._index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self._index < len(self._books):
            result = self._books[self._index]
            self._index += 1
            return result
        else:
            raise StopIteration

class LibraryYearIterator:
    def __init__(self, books, year):
        self._books = [book for book in books if book.year == year]
        self._index = 0

    def __iter__(self):
        return self

    def __next__(self):
        if self._index < len(self._books):
            result = self._books[self._index]
            self._index += 1
            return result
        else:
            raise StopIteration

if __name__ == "__main__":
    library = Library()
    library.add_book(Book("The Great Gatsby", "F. Scott Fitzgerald", 1925))
    library.add_book(Book("1984", "George Orwell", 1949))
    library.add_book(Book("To Kill a Mockingbird", "Harper Lee", 1960))
    library.add_book(Book("Animal Farm", "George Orwell", 1945))

    print("All books in the library:")
    for book in library:
        print(book)

    print("\nBooks by George Orwell:")
    for book in library.get_books_by_author("George Orwell"):
        print(book)

    print("\nBooks published in 1945:")
    for book in library.get_books_by_year(1945):
        print(book)

File: design_patterns/07_Iterator/books.js
class Book {
  constructor(title, author, year) {
    this.title = title;
    this.author = author;
    this.year = year;
  }

  toString() {
    return `${this.title} by ${this.author}, ${this.year}`;
  }
}

class Library {
  constructor() {
    this.books = [];
  }

  addBook(book) {
    this.books.push(book);
  }

  [Symbol.iterator]() {
    return new LibraryIterator(this.books);
  }

  getBooksByAuthor(author) {
    return new LibraryAuthorIterator(this.books, author);
  }

  getBooksByYear(year) {
    return new LibraryYearIterator(this.books, year);
  }
}

class LibraryIterator {
  constructor(books) {
    this._books = books;
    this._index = 0;
  }

  next() {
    if (this._index < this._books.length) {
      return { value: this._books[this._index++], done: false };
    } else {
      return { done: true };
    }
  }
}

class LibraryAuthorIterator {
  constructor(books, author) {
    this._books = books.filter((book) => book.author === author);
    this._index = 0;
  }

  [Symbol.iterator]() {
    return this;
  }

  next() {
    if (this._index < this._books.length) {
      return { value: this._books[this._index++], done: false };
    } else {
      return { done: true };
    }
  }
}

class LibraryYearIterator {
  constructor(books, year) {
    this._books = books.filter((book) => book.year === year);
    this._index = 0;
  }

  [Symbol.iterator]() {
    return this;
  }

  next() {
    if (this._index < this._books.length) {
      return { value: this._books[this._index++], done: false };
    } else {
      return { done: true };
    }
  }
}

// Usage example
const library = new Library();

library.addBook(new Book('The Great Gatsby', 'F. Scott Fitzgerald', 1925));
library.addBook(new Book('1984', 'George Orwell', 1949));
library.addBook(new Book('To Kill a Mockingbird', 'Harper Lee', 1960));
library.addBook(new Book('Animal Farm', 'George Orwell', 1945));

console.log('All books in the library:');
// of => automatically calls next() for you, done flag as well! 
for (const book of library) {  // for loop is sequential 
  console.log(book.toString());
}

// How will you simulate pause? next() by yourself

console.log('\nBooks by George Orwell:');
for (const book of library.getBooksByAuthor('George Orwell')) {
  console.log(book.toString());
}

console.log('\nBooks published in 1945:');
for (const book of library.getBooksByYear(1945)) {
  console.log(book.toString());
}

// const iterator = library[Symbol.iterator]();

// console.log(iterator.next().value.toString())
// console.log(iterator.next().value.toString())
// console.log(iterator.next())
